### PR TITLE
Advanced usage of HTMX

### PR DESCRIPTION
**Pull request by pmlynarek**

### CODE REVIEW

File: app/core/templates/advanced_htmx/partials/car_create.html [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+    <label class="block mb-2 text-sm font-medium text-gray-900 dark:text-white" for="photo">Car Photo</label>
+    <input class="block w-full text-sm text-gray-900 border border-gray-300 rounded-lg cursor-pointer bg-gray-50 dark:text-gray-400 focus:outline-none dark:bg-gray-700 dark:border-gray-600 dark:placeholder-gray-400" id="photo" name="photo" type="file">
+    {% if form.errors.photo %}
+      {% for error in form.errors.photo %}
+        <p class="mt-2 text-sm text-red-600 dark:text-red-500">
+          {{ error }}
+        </p>
+      {% endfor %}
+    {% endif %}
+  </div>
```
Comment (haxoza) [inline app/core/templates/advanced_htmx/partials/car_create.html]: Kind of "partial" template for rendering a simple form field along with label, input and error would be cool!

File: app/core/templates/advanced_htmx/base.html [hunk 2]
```diff
@@ -0,0 +1,25 @@
+{% load static %}
```
Comment (haxoza) [inline app/core/templates/advanced_htmx/base.html]: Destination of this file is somewhat strange.

I would either move all app templates to a standalone directory named `templates` in a project root (see option 2 [here](https://learndjango.com/tutorials/template-structure)), or move it from `core` app to `advanced_htmx` app.

However for the django project that is not a reusable directory I much more prefer the first approach as it makes all templates easily discoverable and browsable in a single directory tree.

File: app/advanced_htmx/ajax.py [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+class TableProductsListAjax(ListView):
+    template_name = "advanced_htmx/partials/table_products_list.html"
+
+    model = Product
+    queryset = Product.objects.select_related("category").order_by("id")
+    paginate_by = 10
+
+    def get(self, request, *args, **kwargs) -> TemplateResponse:
+        response: TemplateResponse = super().get(request, *args, **kwargs)
+        response["HX-Trigger"] = "products-list-page-changed"
```
Comment (haxoza) [inline app/advanced_htmx/ajax.py]: Once we have an enum for that, shouldn't we use its constant value here?

File: app/advanced_htmx/ajax.py [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+    template_name = "advanced_htmx/partials/chat_messages_form.html"
+    http_method_names = ["post"]
+    form_class = ChatMessageForm
+
+    def form_valid(self, form):
+        message: ChatMessage = ChatMessage(
+            signature=form.cleaned_data["signature"],
+            text=form.cleaned_data["text"],
+        )
+        ChatHandler.save_message(message=message)
```
Comment (haxoza) [inline app/advanced_htmx/ajax.py:74]: Love this mapping to dataclass that decouples from django forms ❤️

**Review by haxoza — APPROVED** [review]

**Comment by haxoza** [comment]
When I try out the app myself I get the 500 error on [the car create page](http://localhost:8000/advanced/car).

```
AttributeError: 'CarCreateAjax' object has no attribute 'object'
myproject          | HTTP POST /advanced/ajax/car 500 [0.05, 192.168.65.1:53172]
```

Also, maybe let's change fetching random product from every 1s to 10s or more, because it's polluting the console a bit when you try to follow the logs to see what's happening.

File: app/core/templates/advanced_htmx/partials/car_create.html (hunk 1 above)
Comment (pmlynarek) [reply app/core/templates/advanced_htmx/partials/car_create.html]: Good idea!

I've changed this:
```
  <div class="mb-6">
    <label for="name" class="block mb-2 text-sm font-medium text-gray-900">Car name</label>
    <input
      type="text"
      id="name"
      name="name"
      class="bg-gray-50 border border-gray-300 text-gray-900 text-sm rounded-lg focus:ring-blue-500 focus:border-blue-500 block w-full p-2.5"
      placeholder="Car"
      value="{{ form.name.value }}"
      required
    />
    {% if form.errors.name %}
      {% for error in form.errors.name %}
        <p class="mt-2 text-sm text-red-600">
          {{ error }}
        </p>
      {% endfor %}
    {% endif %}
  </div>
  <div class="mb-6">
    <label for="price" class="block mb-2 text-sm font-medium text-gray-900">Car price</label>
    <input
      type="number"
      step=".01"
      name="price"
      id="price"
      class="bg-gray-50 border border-gray-300 text-gray-900 text-sm rounded-lg focus:ring-blue-500 focus:border-blue-500 block w-full p-2.5"
      value="{{ form.price.value }}"
      required
    />
    {% if form.errors.price %}
      {% for error in form.errors.price %}
        <p class="mt-2 text-sm text-red-600">
          {{ error }}
        </p>
      {% endfor %}
    {% endif %}
  </div>
  <div class="mb-6">
    <label class="block mb-2 text-sm font-medium text-gray-900" for="photo">Car Photo</label>
    <input class="block w-full text-sm text-gray-900 border border-gray-300 rounded-lg cursor-pointer bg-gray-50focus:outline-none" id="photo" name="photo" type="file">
    {% if form.errors.photo %}
      {% for error in form.errors.photo %}
        <p class="mt-2 text-sm text-red-600">
          {{ error }}
        </p>
      {% endfor %}
    {% endif %}
  </div>
```

Into this:
```
  {% comment %} Car name field {% endcomment %}
  {% include "advanced_htmx/mini_partials/car_create_form_field.html" with label_for="name" label="Car name" input_type="text" input_id="name" input_name="name" input_placeholder="Car" input_value=form.name.value input_required="true" errors=form.errors.name %}
  {% comment %} Car price field {% endcomment %}
  {% include "advanced_htmx/mini_partials/car_create_form_field.html" with label_for="price" label="Car price" input_type="number" input_id="price" input_name="price" input_value=form.price.value input_step=".01" input_required="true" errors=form.errors.price %}
  {% comment %} Car photo field {% endcomment %}
  {% include "advanced_htmx/mini_partials/car_create_form_field.html" with label_for="photo" label="Car photo" input_type="file" input_id="photo" input_name="photo" input_class="block w-full text-sm text-gray-900 border border-gray-300 rounded-lg cursor-pointer bg-gray-50focus:outline-none" errors=form.errors.photo %}
```

**Comment by pmlynarek** [comment]
> When I try out the app myself I get the 500 error on [the car create page](http://localhost:8000/advanced/car).
> 
> ```
> AttributeError: 'CarCreateAjax' object has no attribute 'object'
> myproject          | HTTP POST /advanced/ajax/car 500 [0.05, 192.168.65.1:53172]
> ```
> 
> Also, maybe let's change fetching random product from every 1s to 10s or more, because it's polluting the console a bit when you try to follow the logs to see what's happening.

Fixed ✅ For some reasons I had to testing old version of this view and latest changes weren't checked

File: app/advanced_htmx/ajax.py (hunk 3 above)
Comment (pmlynarek) [reply app/advanced_htmx/ajax.py]: Definitely true. I had to miss that

File: app/core/templates/advanced_htmx/base.html (hunk 2 above)
Comment (pmlynarek) [reply app/core/templates/advanced_htmx/base.html]: I will adjust that 🚀
In general my intention was to do sth like Option 2: Project Level

File: app/core/templates/advanced_htmx/partials/car_create.html (hunk 1 above)
Comment (haxoza) [reply app/core/templates/advanced_htmx/partials/car_create.html]: Awesome!

**Pull request merged**